### PR TITLE
Add enter-cloak & exit-cloak effect for Cloak

### LINKED ISSUE #19880: Add an enter-cloak effect and an exit-cloak effect in Cloak
The Cloak in OpenRA is pretty guileless in visual, which may be suitable for the games that a decade ago but I think we can push it forward a little bit. For example, I think add an enter-cloak effect and an exit-cloak effect can make the cloak looks cooler in visual, and  give player in-game information about their cloak type (submarine cloak and stealth-APC cloak). 

And for third party mods they can use this mechanic to make different cloak types for different defector, as well as have better and reasonable visual feedback.

Image from TA
showcase4

Image from SP
A simple flash on cloak effect
GIF

different cloak type
cloak-1

### PR DESCRIPTION
**Pull request by dnqbob**
cloak-showcase

close #19880

### CODE REVIEW

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 1]
```diff
@@ -167,16 +181,34 @@ void ITick.Tick(Actor self)
...
+				}
 			}
 			else if (!isCloaked && wasCloaked)
 			{
 				if (cloakedToken != Actor.InvalidConditionToken)
 					cloakedToken = self.RevokeCondition(cloakedToken);
 
-				if (!(firstTick && Info.InitialDelay == 0) && !otherCloaks.Any(a => a.Cloaked))
-					Game.Sound.Play(SoundType.World, Info.UncloakSound, self.CenterPosition);
+				if (!(firstTick && Info.InitialDelay == 0))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
				if (!firstTick || Info.InitialDelay > 0)
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 2]
```diff
@@ -167,16 +181,34 @@ void ITick.Tick(Actor self)
 					cloakedToken = self.GrantCondition(Info.CloakedCondition);
 
 				// Sounds shouldn't play if the actor starts cloaked
-				if (!(firstTick && Info.InitialDelay == 0) && !otherCloaks.Any(a => a.Cloaked))
+				if (!(firstTick && Info.InitialDelay == 0))
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
				if (!firstTick || Info.InitialDelay > 0)
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 3]
```diff
@@ -88,10 +106,6 @@ public Cloak(CloakInfo info)
 
 		protected override void Created(Actor self)
 		{
-			otherCloaks = self.TraitsImplementing<Cloak>()
-				.Where(c => c != this)
-				.ToArray();
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs:93]: Removing this means that a second cloak sound plays when you have two cloak traits which cloak at the same time/after each other. Have we thought about the original use case here? (It doesn't make sense to me to have two active cloaks, but maybe there is/was a usecase in mind when this was written.)

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 4]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
 		[Desc("The condition to grant to self while cloaked.")]
 		public readonly string CloakedCondition = null;
 
+		[Desc("Which image to use for effect when enter/exit cloak.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		[Desc("Which image to use for the effect played when cloaking or uncloaking.")]
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 5]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+		public readonly string EffectImage = null;
+
+		[Desc("Which sequence to use for effect when enter cloak.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		[Desc("Which effect sequence to play when cloaking.")]
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 6]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+		[SequenceReference(nameof(EffectImage), allowNullImage: true)]
+		public readonly string EnterEffectSequence = null;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		public readonly string CloakEffectSequence = null;
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 7]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+
+		[Desc("Which sequence to use for effect when exit cloak..")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		[Desc("Which effect sequence to play when uncloaking.")]
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 8]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+		[SequenceReference(nameof(EffectImage), allowNullImage: true)]
+		public readonly string ExitEffectSequence = null;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		public readonly string UncloakEffectSequence = null;
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 9]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+
+		[PaletteReference(nameof(IsEffectPlayerPalette))]
+		public readonly string EffectPalette = "effect";
+		public readonly bool IsEffectPlayerPalette = false;
+
+		[Desc("Offset for effect when enter/exit cloak.")]
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		[Desc("Offset for the effect played when cloaking or uncloaking.")]
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 10]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+		[SequenceReference(nameof(EffectImage), allowNullImage: true)]
+		public readonly string EnterEffectSequence = null;
+
+		[Desc("Which sequence to use for effect when exit cloak..")]
+		[SequenceReference(nameof(EffectImage), allowNullImage: true)]
+		public readonly string ExitEffectSequence = null;
+
+		[PaletteReference(nameof(IsEffectPlayerPalette))]
+		public readonly string EffectPalette = "effect";
+		public readonly bool IsEffectPlayerPalette = false;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		public readonly bool EffectPaletteIsPlayerPalette = false;
```

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: Actually this seems unused so please remove it.

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 3 above)
Comment (pchote) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: It was introduced in #9808 which doesn't explain why. There may be related IRC discussion with more context.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: Actually I would suggest use condition control in this use case, instead of hardcode it in order to make only one cloak has its sound.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: Just think about following use case: A submarine is under the effect of Stealth Generator.
Submerge is a kind of cloak, Stealth is a kind of cloak. In regular logic, being stealth doesn't prevent the submarine go under the water surface.

Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: Seems like the reason for this was that Stealth Tanks would cloak when already cloaked by a Stealth Generator. @dnqbob can you please fix this using conditions?

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: Sure. and I also suggest let the Stealth Generator cloak the Stealth Tank when Stealth Tank get damaged heavliy and lose it own stealth ability.

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: Sorry that I am unable to fix those recently due to my laptop's keyboard is dysfunctional and I am trying to fix it.

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 10 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: third-party mod may need this on stealth effect

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 5 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 6 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 7 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 8 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 9 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 10 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fix the spelling, not removed

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 11]
```diff
@@ -63,6 +64,24 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+		[SequenceReference(nameof(EffectImage), allowNullImage: true)]
+		public readonly string CloakEffectSequence = null;
+
+		[Desc("Which effect sequence to play when uncloaking.")]
+		[SequenceReference(nameof(EffectImage), allowNullImage: true)]
+		public readonly string UncloakEffectSequence = null;
+
+		[PaletteReference(nameof(EffectPaletteIsPlayerPalette))]
+		public readonly string EffectPalette = "effect";
+		public readonly bool EffectPaletteIsPlayerPalette = false;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs:83]: This isn't ever used by the code, so if your mod needs this you need to use it too. :)

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 3 above)
Comment (abcdefg30) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: Ok, don't worry, we're not in a rush.

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 11 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:83]: oh damn, sorry for that

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:83]: sparkle

Hope you like those sparkles

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:83]: fixed and I also need to fix TA/RV engine on this :(

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 4 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 3 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs:93]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 12]
```diff
@@ -167,16 +181,42 @@ void ITick.Tick(Actor self)
...
+					Game.Sound.Play(SoundType.World, Info.CloakSound, pos);
+
+					if (Info.EffectImage != null && Info.UncloakEffectSequence != null)
+					{
+						var palette = Info.EffectPalette;
+						if (Info.EffectPaletteIsPlayerPalette)
+							palette += self.Owner.InternalName;
+
+						self.World.AddFrameEndTask(w => w.Add(new SpriteEffect(
+							pos + Info.EffectOffset, w, Info.EffectImage, Info.CloakEffectSequence, palette)));
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
							pos + Info.EffectOffset, w, Info.EffectImage, Info.UncloakEffectSequence, palette)));
```

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 1 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: fixed

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 12 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: :< fixed

File: mods/ts/rules/nod-vehicles.yaml [hunk 13]
```diff
@@ -520,6 +522,7 @@ STNK:
 		IsPlayerPalette: true
 		UncloakOn: Attack, Unload, Infiltrate, Demolish, Damage, Heal
 		PauseOnCondition: cloak-force-disabled || empdisable
+		CloakedCondition: cloak-self
```
Comment (abcdefg30) [inline mods/ts/rules/nod-vehicles.yaml]: Did you test what happens when you build a Stealth Tank within the radius of a Stealth Generator? I would expect it still plays the cloak sound with this setup. Thinking more about it we might need to restore the old checks, as I'm no longer sure this is doable using conditions.

Comment (dnqbob) [reply mods/ts/rules/nod-vehicles.yaml]: Yes, it is still here. going restore the check but I will change it by getting the same cloak type on the unit.

Comment (dnqbob) [reply mods/ts/rules/nod-vehicles.yaml]: ummm, I meet a issue when get the same cloak type. In fact it is cloak type**s**, like targetables. could we make the "InitialDelay = 0" of TS stealth tank's Cloak to avoid this? 

Still the example: submarine is under the effect of Stealth Generator.

Submerge is a kind of cloak, Stealth is a kind of cloak. In regular logic, being stealth and make the sound of steath field doesn't(or may not) prevent submarine make sound of go under the water.

The use case here is more complicated: only one of the two stealth generators can work. In orignal TS, if what I remeber is correct,  the Stealth Tank doesn't recieve the buff from Stealth Generator: when it get heavily damaged, Stealth Generator cannot cloak it like other heavily damage vehicle. If we want to use the gameplay setting like orignal TS it is also simple to solve.

In conclusion, we have 3 ideas here:
1. make InitialDelay=0 for stealth tank cloak (I confirm it is working)
2. change back to TS orignal setting: Stealth Generator doesn't affect Stealth Tank
3. add Cloak something to identify the same cloak method, without using CloakTypes (or we use CloakTypes with a not so confusing rule), which may makes the code more confusing.

Comment (Punsho) [reply mods/ts/rules/nod-vehicles.yaml:525]: I think the most reasonable and logical way for modders is just not having stealth tank be affected by multiple types of stealth via yaml

Comment (Punsho) [reply mods/ts/rules/nod-vehicles.yaml:525]: To me It seems like a bad idea to hardcode the solution

Comment (PunkPun) [reply mods/ts/rules/nod-vehicles.yaml]: If you look at the yaml it makes perfect sense why they cloak twice, so I think it should also be solved via yaml. IMO in this case hardcoding the solution only limits modders

I can see 2 easy ways of fixing this

1. Stealth tank overwrites stealth generator stealth instead of being separate
2. Stealth generator stealth is disabled on stealth tank

Comment (PunkPun) [reply mods/ts/rules/nod-vehicles.yaml]: And one feature creep way. Add a property on cloak called `TriggerCloakOnConditions`. That way you could disable external cloak on stealth tank, but add the external condition to stank cloak. The `TriggerCloakOnConditions` would ignore the internal cloak delay and cloak immediately when any of the set conditions `become` true `cloakgenerator, crate-cloak`. I'm not sure if we can set conditions this way though

Comment (abcdefg30) [reply mods/ts/rules/nod-vehicles.yaml]: What is the problem with using CloakTypes?

Comment (dnqbob) [reply mods/ts/rules/nod-vehicles.yaml]: > What is the problem with using CloakTypes?

CloakTypes cannot be used as a standard that tell OpenRA some Cloak traits belong to one kind of cloak, which have the same cloak&uncloak sound/visual effect and have some special rule when playing the sound/visual effect. For example, Stealth Tank cloak and Cloak Generator cloak.

if we want to solve this problem perfectly, we need another property in CloakInfo like CloakClass/CloakKind or something, restore the check but this check is only effective on the same CloakClass/CloakKind. 

To make it more readable I want to rename CloakTypes into DetectableTypes, and make another CloakType to differ the same cloak.

Comment (abcdefg30) [reply mods/ts/rules/nod-vehicles.yaml]: > To make it more readable I want to rename CloakTypes into DetectableTypes, and make another CloakType to differ the same cloak.

That sounds like a good idea. I suggest `DetectionTypes` (because they're used for detection, and the types themselves are not detectable).

Comment (dnqbob) [reply mods/ts/rules/nod-vehicles.yaml]: And it is done

**Review by abcdefg30 — COMMENTED** [review]
Untested but looks good now in theory. Can you add an update rule for the `CloakTypes` -> `DetectionTypes` changes?

File: mods/ts/rules/nod-vehicles.yaml (hunk 13 above)
Comment (dnqbob) [reply mods/ts/rules/nod-vehicles.yaml]: > Untested but looks good now in theory. Can you add an update rule for the `CloakTypes` -> `DetectionTypes` changes?

I'd love to but I never did the same thing before so could you tell me how to use the rule system?

**Comment by dnqbob** [comment]
> Untested but looks good now in theory. Can you add an update rule for the `CloakTypes` -> `DetectionTypes` changes?

I'd love to but I never did the same thing before so could you tell me how to use the rule system?

**Comment by abcdefg30** [comment]
https://github.com/OpenRA/OpenRA/pull/19867/commits/1fab49ea3b9b062464e639aa42211dc894020ac2 is an example for how to do an update rule:
Create a new file called `RenameCloakTypes.cs` in the `OpenRA.Mods.Common/UpdateRules/Rules/20201213` folder (20201213 because that was the first playtest after release-20200503). Then add an instance of that class to the `// Bleed only changes here` update path in `UpdatePath.cs`.
The content of the file can be similar to my example: Search for each `Cloak` trait, find the `LastChildMatching` `CloakTypes` and rename that node (there is a rename function). Then do the same for all `DetectCloaked` traits.

**Comment by dnqbob** [comment]
thanks, idk if now it is correct

**Comment by dnqbob** [comment]
Change the ` !(firstTick && Info.InitialDelay == 0) && (otherCloaks == null || !otherCloaks.Any(a => a.Cloaked)) ` into `(otherCloaks == null || !otherCloaks.Any(a => a.Cloaked)) && !(firstTick && Info.InitialDelay == 0)`, in the hoping that more performance can be saved

Update: ummm acutally not

**Comment by abcdefg30** [comment]
> Update: ummm acutally not

Can you change it back? (Calling the `Any` before the rest will likely be less performant fwiw, although it won't matter much in practice.)

**Comment by dnqbob** [comment]
it is back now

**Review by abcdefg30 — DISMISSED** [review]
(You can remove the test cases from RA now; before we accidentally merge them. 😅)

**Comment by dnqbob** [comment]
> (You can remove the test cases from RA now; before we accidentally merge them. 😅)


Which one, both?
I kinda want to keep them :(

**Comment by Inq8** [comment]
Can this be expanded to grant a condition when cloaking and a condition when uncloaking?

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 14]
```diff
@@ -57,12 +58,33 @@ public class CloakInfo : PausableConditionalTraitInfo
...
 		public readonly bool IsPlayerPalette = false;
 
-		public readonly BitSet<CloakType> CloakTypes = new BitSet<CloakType>("Cloak");
+		public readonly BitSet<DetectionType> DetectionTypes = new BitSet<DetectionType>("Cloak");
 
 		[GrantedConditionReference]
 		[Desc("The condition to grant to self while cloaked.")]
 		public readonly string CloakedCondition = null;
 
+		[Desc("The type of cloak. Same type of cloaks won't tigger cloaking and uncloaking sound and effect.")]
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cloak.cs]: `tigger`

**Review by pchote — COMMENTED** [review]
One typo I spotted, not a proper review.

I second the request to remove the testcases, IMO the submarine splash in particular doesn't look good and may be confusing for players thinking that something is firing on it.

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 14 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: How about the other usecase of stealth apc?

Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: Fixed and dropped

**Review by reaperrr — DISMISSED** [review]
Ok, though I'm not 100% sure about the RA STNK sparkle effect

**Comment by reaperrr** [comment]
Not merging yet since I don't know how others feel about the STNK sparkle effect.

**Comment by dnqbob** [comment]
> Can this be expanded to grant a condition when cloaking and a condition when uncloaking?

Maybe not in this branch, but it is possible to make it grant a condition when cloaking and a condition when uncloaking

**Comment by Smittytron** [comment]
The animation doesn't follow a moving actor, is this intended?

**Comment by abcdefg30** [comment]
> How about the other usecase of stealth apc?

There hasn't been great support for this change on Discord, so I'd prefer if we can fix the animation not following actors and then remove the effect from the Phase Transport.

**Comment by dnqbob** [comment]
> The animation doesn't follow a moving actor, is this intended?

Yes, considering usecase of submarine (appear and disappear with a water splash) and ninja (disappear with a smoke screen), the cloak effect of those should not follow the actor, but play at the effect triggered position instead.

However, now I make the user can set the follow or not.

>There hasn't been great support for this change on Discord, so I'd prefer if we can fix the animation not following actors and then remove the effect from the Phase Transport.

Spark still not remove for you to test out and test the setting of follow/not follow function. Once it is done I will remove the testcase here.

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 15]
```diff
@@ -57,12 +58,36 @@ public class CloakInfo : PausableConditionalTraitInfo
...
+
+		[PaletteReference(nameof(EffectPaletteIsPlayerPalette))]
+		public readonly string EffectPalette = "effect";
+		public readonly bool EffectPaletteIsPlayerPalette = false;
+
+		[Desc("Offset for the effect played when cloaking or uncloaking.")]
+		public readonly WVec EffectOffset = WVec.Zero;
+
+		[Desc("Should the effect track the actor.")]
+		public readonly bool EffectTrackActor = true;
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]:
```suggestion
		public readonly bool EffectTracksActor = true;
```

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 16]
```diff
@@ -167,16 +195,50 @@ void ITick.Tick(Actor self)
...
+						posfunc = () => pos + Info.EffectOffset;
+
+					if (Info.EffectImage != null && Info.CloakEffectSequence != null)
+					{
+						var palette = Info.EffectPalette;
+						if (Info.EffectPaletteIsPlayerPalette)
+							palette += self.Owner.InternalName;
+
+						self.World.AddFrameEndTask(w => w.Add(new SpriteEffect(
+							posFunc: posfunc, facingFunc: () => WAngle.Zero, w, Info.EffectImage, Info.CloakEffectSequence, palette)));
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Cloak.cs]: Naming the parameters should not be necessary, or is it?
```suggestion
							posfunc, () => WAngle.Zero, w, Info.EffectImage, Info.CloakEffectSequence, palette)));
```

**Review by abcdefg30 — DISMISSED** [review]
Tracking works. ✅

File: OpenRA.Mods.Common/Traits/Cloak.cs (hunk 16 above)
Comment (dnqbob) [reply OpenRA.Mods.Common/Traits/Cloak.cs]: Should be all fixed here and drop the test case

**Review by abcdefg30 — APPROVED** [review]

**Comment by dnqbob** [comment]
Still I hope there can be somewhere use the cloak effect, for maintaining unused code will be hard without an usecase

**Comment by Mailaender** [comment]
[Changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/1579c4bbd93dda9e2ca251ef2ec70c2a4ebcea5f...e43529b5c428a08a81ab5288b036fed5ccf1e9c3)
